flink-runtime/src/main/java/org/apache/flink/runtime/metrics/dump/MetricQueryService.java

```diff
@@ -73,60 +71,50 @@ public String filterCharacters(String input) {
 
 	private final long messageSizeLimit;
 
-	public MetricQueryService(long messageSizeLimit) {
+	public MetricQueryService(RpcService rpcService, String endpointId, long messageSizeLimit) {
+		super(rpcService, endpointId);
 		this.messageSizeLimit = messageSizeLimit;
 	}
 
 	@Override
-	public void postStop() {
+	public CompletableFuture<Void> onStop() {
 		serializer.close();
+		return CompletableFuture.completedFuture(null);
 	}
 
-	@Override
-	public void onReceive(Object message) {
-		try {
-			if (message instanceof AddMetric) {
-				AddMetric added = (AddMetric) message;
-
-				String metricName = added.metricName;
-				Metric metric = added.metric;
-				AbstractMetricGroup group = added.group;
-
-				QueryScopeInfo info = group.getQueryServiceMetricInfo(FILTER);
-
-				if (metric instanceof Counter) {
-					counters.put((Counter) metric, new Tuple2<>(info, FILTER.filterCharacters(metricName)));
-				} else if (metric instanceof Gauge) {
-					gauges.put((Gauge<?>) metric, new Tuple2<>(info, FILTER.filterCharacters(metricName)));
-				} else if (metric instanceof Histogram) {
-					histograms.put((Histogram) metric, new Tuple2<>(info, FILTER.filterCharacters(metricName)));
-				} else if (metric instanceof Meter) {
-					meters.put((Meter) metric, new Tuple2<>(info, FILTER.filterCharacters(metricName)));
-				}
-			} else if (message instanceof RemoveMetric) {
-				Metric metric = (((RemoveMetric) message).metric);
-				if (metric instanceof Counter) {
-					this.counters.remove(metric);
-				} else if (metric instanceof Gauge) {
-					this.gauges.remove(metric);
-				} else if (metric instanceof Histogram) {
-					this.histograms.remove(metric);
-				} else if (metric instanceof Meter) {
-					this.meters.remove(metric);
-				}
-			} else if (message instanceof CreateDump) {
-				MetricDumpSerialization.MetricSerializationResult dump = serializer.serialize(counters, gauges, histograms, meters);
-
-				dump = enforceSizeLimit(dump);
-
-				getSender().tell(dump, getSelf());
-			} else {
-				LOG.warn("MetricQueryServiceActor received an invalid message. " + message.toString());
-				getSender().tell(new Status.Failure(new IOException("MetricQueryServiceActor received an invalid message. " + message.toString())), getSelf());
+	public void addMetric(String metricName, Metric metric, AbstractMetricGroup group) {
+		runAsync(() -> {
```

Comment:
You're right that since add actual calls to `addMetric` are `RpcInvocation`, this method should be already always running in the main thread executor.